File: design-patterns/ts/template-method/transport/bike.ts
// bike.ts

import Transport from "./transport";

class Bike extends Transport {
    createBody(): void {
        console.log("Creating Bike Body");
    }

    addEngine(): void {
        console.log("Adding Engine to Bike");
    }

    addWheel(): void {
        console.log("Adding 2 Wheels to Bike");
    }

    addWing(): void {
        // not required for Bike
    }
}

export default Bike;

File: design-patterns/go/template_method/transport/bike.go
// bike.go

package main

import "fmt"

type Bike struct {
	*Transport
}

func NewBike() (bike *Bike) {
	bike = &Bike{}
	bike.Transport = NewTransport(bike)
	return
}

func (bike *Bike) AddEngine() {
	fmt.Println("Adding Engine to Bike")
}

func (bike *Bike) AddWheel() {
	fmt.Println("Adding 2 Wheels to Bike")
}

func (bike *Bike) AddWing() {
	// Implementation not required for bike
}

func (bike *Bike) CreateBody() {
	fmt.Println("Creating Bike Body")
}
